include/pybind11/eigen/tensor.h

```diff
@@ -0,0 +1,441 @@
+                break;
+
+            case return_value_policy::take_ownership:
+                if (std::is_const<C>::value) {
+                    pybind11_fail("Cannot take ownership of a const reference");
+                }
+
+                parent_object
+                    = capsule(src, [](void *ptr) { Helper::free(reinterpret_cast<Type *>(ptr)); });
+                writeable = true;
+                break;
+
+            case return_value_policy::copy:
+#if defined(__clang_major__) && __clang_major__ <= 3
+                // Hack to work around clang bugs
+                { parent_object = {}; }
+#else
+                parent_object = {};
+#endif
+
+                writeable = true;
+                break;
+
+            case return_value_policy::reference:
+                parent_object = none();
+                writeable = !std::is_const<C>::value;
+                break;
+
+            case return_value_policy::reference_internal:
+                // Default should do the right thing
+                if (!parent) {
+                    pybind11_fail("Cannot use reference internal when there is no parent");
+                }
+                parent_object = reinterpret_borrow<object>(parent);
+                writeable = !std::is_const<C>::value;
+                break;
+
+            default:
+                pybind11_fail("pybind11 bug in eigen.h, please file a bug report");
+        }
+
+        auto result = array_t<typename Type::Scalar, compute_array_flag_from_tensor<Type>()>(
+            Helper::get_shape(*src), src->data(), parent_object);
+
+        if (!writeable) {
+            array_proxy(result.ptr())->flags &= ~detail::npy_api::NPY_ARRAY_WRITEABLE_;
+        }
+
+        return result.release();
+    }
+};
+
+template <typename Type, typename S = enable_if_t<std::is_const<Type>::value>>
+const void *get_array_data_for_type(array &arr) {
+    return arr.data();
+}
+
+template <typename Type, typename S = enable_if_t<!std::is_const<Type>::value>>
+void *get_array_data_for_type(array &arr) {
+    return arr.mutable_data();
+}
+
+template <typename Type, int Options>
+struct type_caster<Eigen::TensorMap<Type, Options>,
+                   typename eigen_tensor_helper<Type>::ValidType> {
+    using MapType = Eigen::TensorMap<Type, Options>;
+    using Helper = eigen_tensor_helper<Type>;
+
+    bool load(handle src, bool /*convert*/) {
+        // Note that we have a lot more checks here as we want to make sure to avoid copies
+        auto arr = reinterpret_borrow<array>(src);
+        if ((arr.flags() & compute_array_flag_from_tensor<Type>()) == 0) {
+            return false;
+        }
+
+        if (!arr.dtype().is(dtype::of<typename Type::Scalar>())) {
+            return false;
+        }
+
+        if (arr.ndim() != Type::NumIndices) {
+            return false;
+        }
+
+        // Use temporary to avoid MSVC warning ...
+        bool is_aligned = (Options & Eigen::Aligned) != 0;
```

Comment:
```suggestion
        constexpr bool is_aligned = (Options & Eigen::Aligned) != 0;
```

Reply:
I had to revert this actually. Switching to constexpr triggers a warning on MSVC.